### PR TITLE
Fix electron app

### LINKED ISSUE #468: Upgrade electron and fix CI builds
As the title implies: the goal here is to upgrade electron version to 20+ and fix CI builds.

Some work for this has already been done, but it's not finished:

- https://github.com/actualbudget/actual/pull/462 (the most recent attempt that got the closest to being done)
- https://github.com/actualbudget/actual/pull/211
- https://github.com/actualbudget/actual/pull/45

Doing this task will solve the following issues:
- https://github.com/actualbudget/actual/issues/44
- https://github.com/actualbudget/actual/issues/293

Once this task is done we can start preparing electron releases too (although in an un-signed format).

### PR DESCRIPTION
**Pull request by Shazib**
Updates to the latest version of electron and moves the backend-frontend communication from node-ipc to websockets. This resolves the previous roadblock regarding `nodeIntegration` .

Done

- Remove node-ipc in favour of websockets. 
- Move file copying out of `preload.js` to avoid importing module `fs` there
- Bump all electron pacakge versions to the latest
- Added new package for finding open ports as node-ipc is gone
- Tweaked webpack config for above changes

Partially fixes #468

Questions/ Pending:
- Literally every single test fails for me, presumably some issue with my setup/environment.
- The websocket communication is not using TLS. I'm not sure how to enable this, or if we even need to as its all local.
- Still need to create the CI for building/deploying but I'm not sure where start in this regard as i have no exp with it. Presumably we will need to point the electron auto-updater to the github releases url's. If people are happy with this PR I will look at adding the CI before its merged.
- In dev mode only, I have disabled TLS security becuase my docker container's cert is not signed. I _assume_ this will be true for other people who spin up the server on thier own hardware. Perhaps I just need to change my cert to one from letsencrypt or something...

Notes.
I have not touched javascript in eons so my apologies if the commit trail is a bit fragmented. I tried to keep them fairly contained and then there is a slightly gnarly final commit fixing all the linter issues... Please let me know if you want me to squash some commits etc.

I initially tried to move this to web workers the same way the web app does it but this was unsuccessful. I have found no way to spin up a worker in one place (frontend/backend) and then pass this worker to the other. The electron ipc channels don't allow you to directly pass objects such as workers, everything is cloned/serialised. Passing a port number so the other end can spin up its own socket works fine.

### CODE REVIEW

**Comment by j-f1** [comment]
Thank you for taking this on!

> Literally every single test fails for me, presumably some issue with my setup/environment.

Running the tests in CI now, so we can see which it is!

> The websocket communication is not using TLS. I'm not sure how to enable this, or if we even need to as its all local.

If it’s entirely to localhost, I don’t think we need to worry too much since any attacker that could sniff the traffic could probably just read the budget files off the disk.

> Still need to create the CI for building/deploying but I'm not sure where start in this regard as i have no exp with it. Presumably we will need to point the electron auto-updater to the github releases url's. If people are happy with this PR I will look at adding the CI before its merged.

I don’t know about deployment but there is a build workflow in `.github/workflows/build.yml` where the Electron part is commented out. Feel free to re-enable it and set it up to run on PRs.

> In dev mode only, I have disabled TLS security becuase my docker container's cert is not signed. I assume this will be true for other people who spin up the server on thier own hardware. Perhaps I just need to change my cert to one from letsencrypt or something...

You don’t need to do this in this PR but it would be great to listen for the [`certificate-error`](https://www.electronjs.org/docs/latest/api/app#event-certificate-error) event and prompt the user to accept it.

> I tried to keep them fairly contained and then there is a slightly gnarly final commit fixing all the linter issues... Please let me know if you want me to squash some commits etc.

We squash as part of the merge process, feel free to make as few or as many commits as you want :)

> I initially tried to move this to web workers the same way the web app does it but this was unsuccessful. I have found no way to spin up a worker in one place (frontend/backend) and then pass this worker to the other. The electron ipc channels don't allow you to directly pass objects such as workers, everything is cloned/serialised. Passing a port number so the other end can spin up its own socket works fine.

That sounds like a reasonable solution.

File: package.json [hunk 1]
```diff
@@ -56,5 +56,8 @@
   "browserslist": [
     "electron 12.0",
     "defaults"
-  ]
+  ],
+  "dependencies": {
+    "get-port-please": "^3.0.1"
```
Comment (j-f1) [inline package.json:61]: Can you move this to the sub package that depends on it?

File: packages/desktop-electron/findOpenSocket.js [hunk 2]
```diff
@@ -1,33 +1,7 @@
...
-        console.log('connected');
-        ipc.disconnect(name);
-        resolve(true);
-      });
-    });
-  });
-}
+const { getRandomPort } = require('get-port-please');
 
 async function findOpenSocket() {
```
Comment (j-f1) [inline packages/desktop-electron/findOpenSocket.js:3]: Since this function is so simple now, can you refactor all of its call sites to call `getRandomPort` directly? (Or do you have a specific reason for keeping it around?)

File: packages/desktop-electron/preload.js [hunk 3]
```diff
@@ -23,19 +20,7 @@ contextBridge.exposeInMainWorld('Actual', {
   },
 
   ipcConnect: (id, func) => {
```
Comment (j-f1) [inline packages/desktop-electron/preload.js:22]: If this does nothing can it be removed?

File: packages/loot-core/src/platform/client/fetch/index.web.ts [hunk 4]
```diff
@@ -9,9 +9,11 @@ let messageQueue = [];
 let socketClient = null;
 
 function connectSocket(name, onOpen) {
```
Comment (j-f1) [inline packages/loot-core/src/platform/client/fetch/index.web.ts:11]: `name` should be `port` now, right?

File: packages/loot-core/src/platform/server/connection/index.d.ts [hunk 5]
```diff
@@ -1,7 +1,7 @@
 import type { ServerEvents } from '../../../types/server-events';
 
 export function init(
-  channel: Window | string,
+  channel: Window | string | int,
```
Comment (j-f1) [inline packages/loot-core/src/platform/server/connection/index.d.ts:4]: TS doesn't have int/float types. Also, is the `string` case used anymore?

```suggestion
  channel: Window | string | number,
```

File: packages/loot-core/webpack/webpack.desktop.config.js [hunk 6]
```diff
@@ -24,14 +24,16 @@ module.exports = {
       '.ts',
       '.tsx',
       '.json',
+      'pegjs',
```
Comment (j-f1) [inline packages/loot-core/webpack/webpack.desktop.config.js:27]: Why was this added?

File: packages/desktop-electron/index.js [hunk 7]
```diff
@@ -92,28 +99,6 @@ function createBackgroundProcess(socketName) {
   });
 }
 
-function createBackgroundWindow(socketName) {
```
Comment (j-f1) [inline packages/desktop-electron/index.js:95]: Is it still possible to open a devtools window for the server process? It can be helpful to be able to debug it sometimes.

**Review by j-f1 — COMMENTED** [review]
A few questions/suggestions. Feel free to ignore if you’re still working on things!

File: packages/desktop-electron/index.js (hunk 7 above)
Comment (Shazib) [reply packages/desktop-electron/index.js:95]: I can in vscode, but if you're running in the terminal you will find it difficult.

The issue was that this background window didnt actually have a toolbar, so you couldn't click `View -> Open Developer Tools`, and the ctrl+shift+i shortcut didnt work either, so you didn't actually have devtools for the server process anyway. 

I tried forcing the window to have a toolbar but this removed the toolbar from the main window. Someone with more electron experience might be able to solve this.

File: packages/loot-core/webpack/webpack.desktop.config.js (hunk 6 above)
Comment (Shazib) [reply packages/loot-core/webpack/webpack.desktop.config.js:27]: It was the only way i could get it to build - goal templates use some .pegjs files that I guess were not getting included in the webpack bundle file. 
If you see [this](https://github.com/actualbudget/actual/pull/988/commits/73a4c55c6ecddc37bd7d98f34b193448cc99e4af) commit by Matiss you can see a to-do item referencing this issue.

**Comment by j-f1** [comment]
Can you re-run `yarn`? It looks like all of the builds are unhappy with the yarn.lock file.

**Comment by Shazib** [comment]
> ~Can you re-run `yarn`? It looks like all of the builds are unhappy with the yarn.lock file.~ never mind, changing a few other things on my end and will pick that up too

Sorry already done

**Review by j-f1 — APPROVED** [review]
This looks good to me as a starting point but there’s definitely a bunch we need to do before releasing. I’m personally happy to merge as is though!

**Comment by Shazib** [comment]
> This looks good to me as a starting point but there’s definitely a bunch we need to do before releasing. I’m personally happy to merge as is though!

Thanks! Let me know if you want me to tackle anything w.r.t release process/workflows etc. 

I suppose it may need/want its own versioning/release list?

**Comment by j-f1** [comment]
In terms of versioning, I think we would want to release the Electron app at the same time we release the `@actual-app/web` package, and we’d want to match versions between the web and Electron UIs.

For the release process, we’d need to figure out code signing. I can help with macOS since I have a developer account at the moment, but may not be able to use it after a few months. I don’t know what Windows would look like.

There are also a few issues I saw giving the built Electron app from GH Actions a quick spin:

- ~I can’t open the devtools for some reason in the production app.~ it seems like it does open, it just takes a really long time?
- The server picker screen shows `app://actual` as a default server option. I don’t think this makes sense, we should hide that button in Electron.
- I get this error when trying to open any budget file:

    ```
	[Exception] Error: The module '/var/folders/r_/3vv7gfs93tg9xq8dfr83lfpm0000gn/T/.org.chromium.Chromium.sddyDN'
	was compiled against a different Node.js version using
	NODE_MODULE_VERSION 93. This version of Node.js requires
	NODE_MODULE_VERSION 114. Please try re-compiling or re-installing
	the module (for instance, using `npm rebuild` or `npm install`).
	    at process.func [as dlopen] (node:electron/js2c/asar_bundle:2:1822)
	    at Module._extensions..node (node:internal/modules/cjs/loader:1326:18)
	    at Object.func [as .node] (node:electron/js2c/asar_bundle:2:2049)
	    at Module.load (node:internal/modules/cjs/loader:1096:32)
	    at Module._load (node:internal/modules/cjs/loader:937:12)
	    at f._load (node:electron/js2c/asar_bundle:2:13330)
	    at Module.require (node:internal/modules/cjs/loader:1120:19)
	    at require (node:internal/modules/cjs/helpers:103:18)
	    at bindings (/Volumes/Actual 22.12.3/Actual.app/Contents/Resources/app.asar/node_modules/bindings/bindings.js:112:48)
	    at new Database (/Volumes/Actual 22.12.3/Actual.app/Contents/Resources/app.asar/node_modules/better-sqlite3/lib/database.js:48:64) {
	  code: 'ERR_DLOPEN_FAILED'
      ```

**Comment by Shazib** [comment]
> ~I can’t open the devtools for some reason in the production app.~ it seems like it does open, it just takes a really long time?
This seems to work fine in windows, no delay.
> * The server picker screen shows `app://actual` as a default server option. I don’t think this makes sense, we should hide that button in Electron.
I will rectify this
> * I get this error when trying to open any budget file:
I have seen this error before and am also seeing it in the windows version.

As this PR is in the approved stage should I consider this one 'locked' and submit further changes in a new PR?

**Comment by j-f1** [comment]
> As this PR is in the approved stage should I consider this one 'locked' and submit further changes in a new PR?

Whichever way you prefer!

**Comment by Shazib** [comment]
> * I get this error when trying to open any budget file:

This error occurs in the dev builds as well and (for me) is rectified by running `yarn rebuild-electron`. Can we update the prod build to do this?

**Comment by j-f1** [comment]
Sure!

**Comment by Shazib** [comment]
> Sure!

The latest build from GH appears to open budgets ok now, and hides that invalid button.

Can't comment on dev tools on mac though

Thanks!

File: packages/loot-core/src/shared/environment.ts [hunk 8]
```diff
@@ -9,3 +9,10 @@ export function isDevelopmentEnvironment() {
 export function isNonProductionEnvironment() {
   return isPreviewEnvironment() || isDevelopmentEnvironment();
 }
+
+export function isElectron() {
```
Comment (j-f1) [inline packages/loot-core/src/shared/environment.ts:13]: I think `isBrowser` in `loot-core/src/client/platform.ts` handles this already — is that missing something?

**Review by j-f1 — APPROVED** [review]
One small comment about how you’re checking for Electron vs web, but this still feels good to merge for me. I’ve spotted a couple of usability issues but I may try fixing them myself once this is merged, or comment them them on a new issue I’ll open to discuss the Electron version once this PR merges.

File: packages/loot-core/src/shared/environment.ts (hunk 8 above)
Comment (Shazib) [reply packages/loot-core/src/shared/environment.ts:13]: Sorry just didnt spot that function existed.

**Comment by MatissJanis** [comment]
How do I launch it?

```
$ yarn start:desktop
Usage Error: Workspace 'Actual' not found. Did you mean any of the following:
  - @actual-app/api
  - @actual-app/import-ynab4
  - @actual-app/import-ynab5
  - @actual-app/web
  - actual
  - desktop-electron
  - eslint-plugin-actual
  - loot-core
  - node-libofx?

$ yarn workspace <workspaceName> <commandName> ...
ERROR: "start:desktop-electron" exited with 1.
```

**Comment by j-f1** [comment]
I think that workspace was recently renamed — try `desktop-electron` instead of `Actual` in the scripts.

**Review by MatissJanis — APPROVED** [review]
Overall LGTM! Thanks for working on Electron.

After this is merged we will still need to work on it before we can do a release though. I'm noticing a bunch of smaller and bigger bugs. For example: the side-nav just doesn't work.

**Comment by Shazib** [comment]
> Overall LGTM! Thanks for working on Electron.

Thanks! I've just updated from master so should be no more conflicts (as of right now!)

> After this is merged we will still need to work on it before we can do a release though. I'm noticing a bunch of smaller and bigger bugs. For example: the side-nav just doesn't work.

I've got these issues on my to-do list! Hopefully another PR on the way once this is merged.

**Comment by j-f1** [comment]
Thank you for getting the ball rolling on this!
